circuit/types/integers/src/mul_checked.rs

```diff
@@ -211,42 +211,52 @@ impl<E: Environment, I: IntegerType> Integer<E, I> {
 }
 
 impl<E: Environment, I: IntegerType> Metrics<dyn MulChecked<Integer<E, I>, Output = Integer<E, I>>> for Integer<E, I> {
-    type Case = (Mode, Mode);
+    type Case = (Mode, Mode, bool, bool);
 
     fn count(case: &Self::Case) -> Count {
         // Case 1 - 2 integers fit in 1 field element (u8, u16, u32, u64, i8, i16, i32, i64).
         if 2 * I::BITS < (E::BaseField::size_in_bits() - 1) as u64 {
             match I::is_signed() {
                 // Signed case
-                true => match (case.0, case.1) {
-                    (Mode::Constant, Mode::Constant) => Count::is(I::BITS, 0, 0, 0),
-                    (Mode::Constant, _) | (_, Mode::Constant) => {
+                true => match (case.0, case.1, case.2, case.3) {
```

Comment:
I think you can revert these code changes (and ignore `case.2, case.3`), right?

Reply:
If we revert this file to what it was before this PR, these tests will fail. Same for your other comments. 

This is because of the way we aggregate `LinearCombination`s. When a `LinearCombination` is multiplied by 0, it will now only have a `constant: 0`, instead of e.g. `terms: [(0*Private(some_variable_index))]`.


For example, before this PR, here we had only one case:
```
(Mode::Constant, _) | (_, Mode::Constant) => {
                        Count::is(4 * I::BITS, 0, (7 * I::BITS) + 4, (8 * I::BITS) + 9)
                    }
```

After this PR, if either value in the test case is zero, we get less constraints and more constants:
```
(Mode::Constant, _, true, _) | (_, Mode::Constant, _, true) => {
                        Count::is(7 * I::BITS, 0, (2 * I::BITS) + 1, (2 * I::BITS) + 2)
                    }
 (Mode::Constant, _, false, _) | (_, Mode::Constant, _, false) => {
                        Count::is(4 * I::BITS, 0, (7 * I::BITS) + 4, (8 * I::BITS) + 9)
                    }
```

As mentioned in Slack, I realized some tests are still unaddressed, I will chat with Pranav and look into it again this week. I'm also going to document the way we aggregate the linear combinations for the auditors.